woocommerce/woocommerce
# Add queue and process endpoints

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes #53803 

This is the first draft, but I’m opening it as it’s blocking other issues. We’ll need to discuss how to detect conflicts and whether or not we want to support asynchronous uploads before coming back to this issue.

- This PR does not check for conflicts. 
- Added `/wp-json/wc-admin/blueprint/queue` and `/wp-json/wc-admin/blueprint/process` endpoints.
- Updated `/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard&step=intro-builder` endpoint to use the new endpoints. 

`/queue` endpoint usage example: https://github.com/woocommerce/woocommerce/blob/update/blueprint-endpoints-for-new-import-flow/plugins/woocommerce/client/admin/client/core-profiler/pages/BuilderIntro.tsx#L53

The endpoint returns the following response:

```
{
    "reference": "string or null if there was an error",
    "process_nonce": "nonce value required by the process endpoint",
    "error_type": "schema_validation | upload | conflicts",
    "errors": []
}
```

`reference` and `process_nonce` must be passed to `process` endpoint.

Example: https://github.com/woocommerce/woocommerce/blob/update/blueprint-endpoints-for-new-import-flow/plugins/woocommerce/client/admin/client/core-profiler/pages/BuilderIntro.tsx#L64






### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch and `blueprint` feature flag.
2. Navigate to `WooCommerce -> Settings -> Advanced -> Blueprint` and export a schema.
3. Naivgate to /wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard&step=intro-builder
4. Upload your schema
5. Inspect network requests and confirm `queue` and `process` endpoints are used.



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>


#### Comment 
This feature is behind a feature flag and not intended for public use yet.

</details>
